### ISSUE TITLE
bikeshed update can't download files

### ISSUE THREAD

**Issue by rtoy** [tabatkins/bikeshed#1801]
Once in a while when I do `git pull --rebase; pip3 install -e .; bikeshed update` I sometimes get (one or more) warnings like this:
```
WARNING: Couldn't download file 'https://raw.githubusercontent.com/tabatkins/bikeshed-data/master/data/wpt-tests.txt'.
Cannot connect to host raw.githubusercontent.com:443 ssl:True [SSLCertVerificationError: (1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1123)')]
```

Don't know if this is a problem with bikeshed or with github being messed up in some way.  But I thought I'd report this anyway. It seems more common these last week or so.  I don't recall ever seeing these before then.

**Comment by tabatkins** [comment]
...huh. I *suspect* this is just a transient network error? That said, I *do* have some retry logic built into the *full* update (querying all of the original data sources); I should probably apply it to the *standard* update process (via manifest, only querying static files on GitHub) as well. ^_^

**Comment by rtoy** [comment]
When I run `bikeshed update` again, bikeshed just says everything is up-to-date (without any other messages) so it's hard to tell if it really is up-to-date or just failed to note that it still couldn't download the files.

**Comment by tabatkins** [comment]
Yeah, if there's a problem downloading, it looks like I'm not reflecting that in the manifest data, so there'll just be no way to tell (until that file updates again, at least). Definitely needs fixing!

**Comment by tabatkins** [comment]
Okay, at least partially fixed. Errors in downloading or saving files from the manifest will be recorded in the manifest, and will trigger a re-download next time you `bikeshed update` even if the remote data hasn't changed.

I still need to add retry logic into the function, but we're in a better spot now at least.

**Comment by rtoy** [comment]
That's fantastic!  At least I know that if bikeshed spec does weird things, it could be because the update didn't work correctly.

I tried now and fails the first time (after doing a git pull; pip install -e .; bikeshed update).  When I update again, I get (roughly?) the same messages.

So, yeah, this is great!  I don't mind too much having to retry manually as long as I know I need to.

**Comment by rtoy** [comment]
In case it matters, one of the messages I get is 
```
WARNING: Couldn't download file 'https://raw.githubusercontent.com/tabatkins/bikeshed-data/master/data/anchors/anchors-ju.data'.
Cannot connect to host raw.githubusercontent.com:443 ssl:True [SSLCertVerificationError: (1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1123)')]
```
but `curl https://raw.githubusercontent.com/tabatkins/bikeshed-data/master/data/anchors/anchors-ju.data` downloads the data just fine.  Maybe my local setup is messed up?  I only have this problem on my mac, not on linux.

**Comment by tabatkins** [comment]
Given that all of the files are downloaded identically and from the same website (same repo, even) but only one of them is triggering problems, yeah, it's some weird local issue. 😃

**Comment by rtoy** [comment]
I'm currently getting 35 failed files now.  I have no idea how to figure out my weird local issue.  Any hints?

**Comment by tabatkins** [comment]
So I just pushed retry logic in, so hopefully these'll fix?

And unfortunately, no, I have absolutely no idea how any of this works. It's basically magic and the fact that it works at all is due solely to other people being clever.

Is it failing for *all* of the files it now tries to update? I *just now* pushed an update to the message so it tells you how many total files it tried to update; is it failing for them all, or just some?

**Comment by rtoy** [comment]
A couple of comments.  I dislike the new error messages which looks like:
```
Updating via manifest...
WARNING: Previous update had some download errors, so re-running...
Updating 47 files...
Updated 0/47, 1 errors...
FATAL ERROR: Done, but there were 47 errors (of 47 total) in downloading or saving. Run `bikeshed update` again to retry.
 ✘  Did not generate, due to fatal errors
```

I don't get any information on what failed.  That makes it hard for me and you to figure out what's failing.

However, having said that, I noticed there's a `--skip-manifest` option.  I tried that, and it took a while to download everything, but it succeeded to download everything.  Then I ran `bikeshed update` which said there's nothing to update. Sweet!

Perhaps my manifest file was messed up somehow?  Anyway, I'm going to try to do an update again tomorrow to see if the issue is gone.

**Comment by tabatkins** [comment]
Oh, duh, sorry, I did indeed strip out the individual error messages. That wasn't an intended change, I'll put those back.

Yeah, `--skip-manifest` does the full "download from sources, assemble all the data manually" steps that the bikeshed-data bot normally does. Because those sources are slow, and the assembling is slow, it's not generally great (plus I wanted to make sure that, as Bikeshed usage went up, I wasn't spamming the source servers with db queries; GH can handle being spammed with static file requests far better).

I'm guessing, just because I think we've run into similar stuff before, that it's something fucked up with Mac's Python HTTPS stack. It's not a *general* https failure, because the manual update *also* does everything over https, so it'd be somethign specific to GH. But some of the `--skip-manifest` sources are *also* just static files on GH anyway, so if all *those* succeeded, I'm really not sure what's up. :(

Let's hope we can figure this out!

**Comment by rtoy** [comment]
Hmm. Maybe it's my python version.  I have 3 potentially different versions of python3 on my mac.  One in /opt/local/bin (macports), /usr/local/bin (don't know where that came from) and /usr/bin, in that order.

I'll try to test tomorrow if this makes a difference.

**Comment by rtoy** [comment]
`git pull; pip3 install; bikeshed update` still gives a few warnings about not downloading files.  And some investigation shows that bikeshed is using python from macports.  Perhaps I should remove everything and start over and use a different python3.

**Comment by frivoal** [comment]
Not entirely sure what was up, but I was recently stuck in a state where bikeshed would have outdated definitions for terminology and tests (and maybe more), and running `bikeshed update` would always report success but leave me in the same out of date state. `bikeshed update --skip-manifest` did fix it though.
Since it is now fixed, I cannot give more details as to what exactly was going on, but I thought collecting anecdotes might be helpful. (I was on the latest bikeshed, pulled from git, with python 3.8.6)

**Comment by zcorpan** [comment]
I've now run into the same issue or something similar. Running `bikeshed update` gives:

```
WARNING: Previous update had some download errors, so re-running...
Updating 1072 files...
WARNING: Error downloading caniuse/feature-wasm-simd.json, full error was:
Cannot connect to host raw.githubusercontent.com:443 ssl:True [SSLCertVerificationError: (1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1081)')]
Updated 0/1072, 1 errors...
(same for all 1072 files)
FATAL ERROR: Done, but there were 1072 errors (of 1072 total) in downloading or saving. Run `bikeshed update` again to retry.
```

Further, after this, my wifi is broken for all devices for some time... Happened yesterday and today immediately after `bikeshed update`. I don't know what's up with *that*.

**Comment by tabatkins** [comment]
That sounds bizarre! Again, uh, no idea what could possibly be happening, I'm just doing a perfectly by-the-books usage of `requests` to fetch some static files from GitHub. Nothing has changed on the Bikeshed side, and it works locally, so the problem is somewhere between you and GitHub. :/
